docs/architecture/architecture.md

```diff
@@ -0,0 +1,671 @@
+  after the `StartPodCmd` has been issued for starting a pod. This command also
+  starts the container process.
+- `ExecCmd`: Executes a new process within an already running container.
+- `KillContainerCmd`: `cc-shim` uses this to send signals to a container process.
+- `WinsizeCmd`: `cc-shim` uses this to change the terminal size of the terminal
+  associated with a container.
+- `RemoveContainerCmd`: Removes a container from the pod. This command will fail
+  if the container is in a running state.
+- `Destroypod`: Removes all containers within a pod . All containers need to be
+  in a stopped state for this command to succeed. The command also frees resources
+  associated with the pod.
+
+Each control message is composed of a command code and a payload required for
+the command:
+
+```
+  ┌────────────────┬────────────────┬──────────────────────────────┐
+  │  Command Code  │     Length     │ Payload(request or response) │
+  │   (32 bits)    │    (32 bits)   │     (data length bytes)      │
+  └────────────────┴────────────────┴──────────────────────────────┘
+```
+- `Command Code` is the predefined id associated with a command.
+- `Length` is the size of the entire message in bytes and encoded in network order.
+- `Payload` is the JSON-encoded command request or response data.
+
+Each stream message is composed of a stream sequence code and a payload containing
+the stream data.
+
+```
+  ┌────────────────┬────────────────┬──────────────────────────────┐
+  │  Sequence Code │     Length     │          Payload             │
+  │   (64 bits)    │    (64 bits)   │     (data length bytes)      │
+  └────────────────┴────────────────┴──────────────────────────────┘
+```
+- `Sequence code` is the 64 bit sequence number assigned by `cc-agent` for a stream.
+- `Length` is the size of the entire stream message in bytes and encoded in network
+  order.
+- `Payload` is the stream data.
+
+The `cc-agent` makes use of [`libcontainer`](https://github.com/opencontainers/runc/tree/master/libcontainer)
+to manage the lifecycle of the container. This way the `cc-agent` reuses most
+of the code used by [`runc`](https://github.com/opencontainers/runc).
+
+## Runtime
+
+`cc-runtime` is an OCI compatible container runtime and is responsible for handling
+all commands specified by
+[the OCI runtime specification](https://github.com/opencontainers/runtime-spec)
+and launching `cc-shim` instances.
+
+`cc-runtime` heavily utilizes the
+[virtcontainers project](https://github.com/containers/virtcontainers), which
+provides a generic, runtime-specification agnostic, hardware-virtualized containers
+library.
+
+### Configuration
+
+The runtime uses a TOML format configuration file called `configuration.toml`. By
+default this file is installed in the `/etc/clear-containers` directory.
+Most users will not need to modify the configuration file.
+
+The file is well commented and provides a few "knobs" that can modify the
+behaviour of the runtime.
```

Comment:
I wrote this but on re-reading, I think it would be better to say, "that can be used to modify the behaviour of the runtime."